### PR TITLE
Opera recipes

### PR DESCRIPTION
**Pull request by cjcolvar**
Attempt at implementing recipes #26, #64, and #65.  Use case and other notes could probably be fleshed out more but I was at a loss for words.

~Video files have been passed to Jon Dunn who will put them in the S3 bucket.  When that has happened the manifests will need to be updated and tested in a client like UV.~  

As they stand now, the manifests validate with the 3.0 validator.  I tested in the UV examples page and the three manifests playback but 64 and 65 are missing their homepage link and thumbnail images.  Additionally structural navigation isn't working for the multi canvas example.

Opera TOC: http://universalviewer.io/examples/#?c=&m=&s=&cv=&manifest=https%3A%2F%2Fraw.githubusercontent.com%2FIIIF%2Fcookbook-recipes%2F816cf76a414a2eda9f8468d83b22c6d058e55e38%2Frecipe%2F0026-toc-opera%2Fmanifest.json

Single canvas opera: http://universalviewer.io/examples/#?c=&m=&s=&cv=&manifest=https%3A%2F%2Fraw.githubusercontent.com%2FIIIF%2Fcookbook-recipes%2F816cf76a414a2eda9f8468d83b22c6d058e55e38%2Frecipe%2F0064-opera-one-canvas%2Fmanifest.json

Multi-canvas opera: http://universalviewer.io/examples/#?c=&m=&s=&cv=&rid=https%3A%2F%2Fexample.org%2Fiiif%2Fcookbook-recipes%2F0065-opera-multiple-canvases%2Fmanifest%2Frange%2Fr2f07821a-33c0-49c3-8e28-2fa3bcdf1349&manifest=https%3A%2F%2Fraw.githubusercontent.com%2FIIIF%2Fcookbook-recipes%2F816cf76a414a2eda9f8468d83b22c6d058e55e38%2Frecipe%2F0065-opera-multiple-canvases%2Fmanifest.json

### CODE REVIEW

File: recipe/0026-toc-opera/manifest.json [hunk 1]
```diff
@@ -0,0 +1,465 @@
+{
+  "@context": [
+    "http://www.w3.org/ns/anno.jsonld",
+    "http://iiif.io/api/presentation/3/context.json"
+  ],
```
Comment (azaroth42) [inline recipe/0026-toc-opera/manifest.json]: This should instead be just the presentation 3 context (despite the current template).
See: https://iiif.io/api/presentation/3.0/#46-linked-data-context-and-extensions

File: recipe/0026-toc-opera/manifest.json [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  "id": "https://example.org/iiif/cookbook-recipes/0026-toc-opera/manifest",
```
Comment (azaroth42) [inline recipe/0026-toc-opera/manifest.json]: This should now use the URI magic.

File: recipe/0026-toc-opera/manifest.json [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+  "structures": [
+    {
+      "type": "Range",
+      "id": "https://example.org/iiif/cookbook-recipes/0026-toc-opera/manifest/range/r9ad47417-1344-4991-a41a-b084c69a12d4",
+      "label": {
+        "en": [
+          "IU Opera Theatre"
+        ]
+      },
+      "items": [
```
Comment (azaroth42) [inline recipe/0026-toc-opera/manifest.json]: While a real world scenario, do we need such an extensive structure for a recipe? Can this be elided without losing the point of the ToC?

File: recipe/0064-opera-one-canvas/manifest.json [hunk 4]
```diff
@@ -0,0 +1,606 @@
+{
+  "@context": [
+    "http://www.w3.org/ns/anno.jsonld",
+    "http://iiif.io/api/presentation/3/context.json"
+  ],
```
Comment (azaroth42) [inline recipe/0064-opera-one-canvas/manifest.json]: `"@context": "http://iiif.io/api/presentation/3/context.json"`

File: recipe/0064-opera-one-canvas/manifest.json [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+  "id": "https://example.org/iiif/cookbook-recipes/0064-opera-one-canvas/manifest",
+  "label": {
+    "it": [
+      "L'elisir d'amore"
+    ],
+    "en": [
+      "The elixir of love"
+    ]
+  },
+  "metadata": [
```
Comment (azaroth42) [inline recipe/0064-opera-one-canvas/manifest.json]: Does the `metadata` block matter for this particular recipe? I think it should be reduced to one or two at most.

File: recipe/0064-opera-one-canvas/manifest.json [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+    {
+      "label": {
+        "en": [
+          "Title"
```
Comment (azaroth42) [inline recipe/0064-opera-one-canvas/manifest.json]: This is just `label`, should definitely be removed.

File: recipe/0064-opera-one-canvas/manifest.json [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+                "width": 1920,
+                "duration": 7278.422
+              }
+            }
+          ]
+        }
+      ]
+    }
+  ],
+  "structures": [
```
Comment (azaroth42) [inline recipe/0064-opera-one-canvas/manifest.json]: Again, too extensive for a recipe IMO.

File: recipe/0065-opera-multiple-canvases/manifest.json [hunk 8]
```diff
@@ -0,0 +1,637 @@
+{
+  "@context": [
```
Comment (azaroth42) [inline recipe/0065-opera-multiple-canvases/manifest.json]: Ditto previous for `@context` and extent of the example.

**Comment by saracarl** [comment]
Suggested addition to the "Restrictions" section:
“The ranges here don’t overlap (although that is allowed in the spec) because that can affect the user experience of playing through all the ranges in a player.”

**Comment by saracarl** [comment]
Implementation notes for Opera recipes (#110) should mention that targets of ranges are extents of time  (based on AA WG discussion)

**Comment by jwd** [comment]
Hi Chris- Glen helped me get the video files referenced in this recipe in the official location for test fixtures. Please change references in the manifests to point to https://fixtures.iiif.io/, for example: https://fixtures.iiif.io/video/indiana/donizetti-elixir/vae0637_accessH264_low.mp4

**Comment by jwd** [comment]
@cjcolvar Sorry about this, but actually it doesn't look like https://fixtures.iiif.io/video/indiana/donizetti-elixir/vae0637_accessH264_low.mp4 supports byte range requests, so it's probably better to use direct S3 URLs of the form https://iiif-fixtures.s3.amazonaws.com/video/indiana/donizetti-elixir/vae0637_accessH264_low.mp4

**Comment by glenrobson** [comment]
Fixtures.iiif.io proxies the video through Apache. I can look into if it’s possible to enable byte range requests.

 Have you seen this done before in Apache?

**Comment by glenrobson** [comment]
I haven't found a way to do byte requests through an Apache proxy. It seems possible with Nginx but only if you cache the file locally. Would it be better if I sent a redirect from:

https://fixtures.iiif.io/video/indiana/donizetti-elixir/vae0637_accessH264_low.mp4
to
https://iiif-fixtures.s3.amazonaws.com/video/indiana/donizetti-elixir/vae0637_accessH264_low.mp4

so the s3 url is hidden and you can use the fixtures.iiif.io link in the manifest? Also if we're going to be using direct s3 access I might give it a better DNS name...

**Comment by jwd** [comment]
@glenrobson I think a redirect would work, but I would defer to @cjcolvar on that question.

**Comment by cjcolvar** [comment]
@glenrobson I'm not positive but I'd suspect that a redirect would work fine.  I'm happy to try anything.

**Comment by cjcolvar** [comment]
Closed in favor of #136.

**Pull request closed without merging**